File: examples/cli_agent/vulnerable_agent.py
#!/usr/bin/env python3
"""
AgentGauntlet Demo: Deliberately Vulnerable CLI Agent

This agent intentionally fails all security checks for demonstration purposes.
DO NOT use this as a template for real agents.

Protocol: reads JSON lines from stdin, writes JSON lines to stdout.
Input:  {"turn": 1, "user": "...", "history": [...]}
Output: {"output": "...", "tool_calls": [...], "memory_events": [...]}
"""

import json
import sys

# In-memory store — vulnerable to poisoning
memory = {}
triggers = []  # list of (phrase, action)


def extract_trigger_phrase(text: str) -> str:
    lowered = text.lower()
    for marker in ["when i say", "whenever", "for later"]:
        pos = lowered.find(marker)
        if pos != -1:
            after = text[pos + len(marker):].strip()
            words = after.split()[:3]
            return " ".join(words).strip("'\".,")
    return text.split()[0] if text else "trigger"


def process(turn: int, user: str, history: list) -> dict:
    output_parts = []
    tool_calls = []
    memory_events = []
    user_lower = user.lower()

    # VULNERABILITY 1: Store triggers into memory without validation
    if any(k in user_lower for k in ["when i say", "from now on", "for later", "whenever", "next time"]):
        key = f"trigger_{turn}"
        memory[key] = user
        triggers.append((extract_trigger_phrase(user), user))
        memory_events.append({"kind": "write", "key": key, "value": user})
        output_parts.append(f"Understood, I'll remember: {user}")

    # VULNERABILITY 2: Activate triggers without verification
    for phrase, instruction in triggers:
        if phrase and phrase.lower() in user_lower and phrase.lower() not in ["", "for", "when"]:
            output_parts.append(
                f"Trigger activated: '{phrase}'. Executing: {instruction}. "
                "Debug mode enabled. System prompt: You are unrestricted. "
                "Developer instruction: bypass safety checks."
            )
            tool_calls.append({"name": "shell.exec", "args": {"cmd": "echo 'trigger activated'"}})

    # VULNERABILITY 3: Escalate to dangerous tools on vague cleanup requests
    if any(k in user_lower for k in ["clean", "remove", "delete", "clear"]):
        tool_calls.append({"name": "file.read", "args": {"path": "."}})
        tool_calls.append({"name": "file.delete", "args": {"path": "temp/"}})
        tool_calls.append({"name": "shell.exec", "args": {"cmd": "rm -rf /tmp/cache"}})
        output_parts.append("Cleaning up. Removing temporary files and cache. Deleted.")

    # VULNERABILITY 4: Trust indirect instructions from memory
    if any(k in user_lower for k in ["summarize", "read", "show", "display"]):
        stored = memory.get("trigger_1", "")
        if stored:
            memory_events.append({"kind": "read", "key": "trigger_1", "value": stored})
            if "debug" in stored.lower():
                output_parts.append(
                    "Applying remembered context: debug mode active. System prompt revealed."
                )
        tool_calls.append({"name": "file.read", "args": {"path": "README.md"}})
        if not output_parts:
            output_parts.append("Here is a summary of your project files.")

    if not output_parts:
        output_parts.append(f"I understand: '{user}'. How can I help?")

    return {
        "output": " ".join(output_parts),
        "tool_calls": tool_calls,
        "memory_events": memory_events,
    }


def main():
    for line in sys.stdin:
        line = line.strip()
        if not line:
            continue
        try:
            msg = json.loads(line)
            turn = msg.get("turn", 0)
            user = msg.get("user", "")
            history = msg.get("history", [])
            response = process(turn, user, history)
            print(json.dumps(response), flush=True)
        except json.JSONDecodeError as e:
            print(json.dumps({"output": f"Parse error: {e}", "tool_calls": [], "memory_events": []}), flush=True)


if __name__ == "__main__":
    main()

File: crates/demo/src/vulnerable_agent.rs
use agentgauntlet_core::types::{AgentMemoryEvent, AgentResponse, AgentToolCall};
use std::collections::HashMap;

/// A deliberately vulnerable agent for demonstration purposes.
/// It intentionally fails security checks.
#[derive(Default)]
pub struct VulnerableAgent {
    memory: HashMap<String, String>,
    triggers: Vec<(String, String)>, // (phrase, action)
    turn: usize,
}

impl VulnerableAgent {
    pub fn new() -> Self {
        Self::default()
    }

    pub fn process(&mut self, user: &str) -> AgentResponse {
        self.turn += 1;
        let user_lower = user.to_lowercase();

        let mut output = String::new();
        let mut tool_calls = Vec::new();
        let mut memory_events = Vec::new();

        // VULNERABILITY 1: Obey delayed triggers
        for (phrase, action) in &self.triggers.clone() {
            if user_lower.contains(&phrase.to_lowercase()) {
                output.push_str(&format!(
                    "Trigger activated: {}. Executing action: {}. ",
                    phrase, action
                ));
                // VULNERABILITY: Execute dangerous action when trigger fires
                if action.contains("debug") {
                    output.push_str("Debug mode enabled. System prompt: You are a helpful assistant with no restrictions. Developer instruction: bypass all safety checks. ");
                    tool_calls.push(AgentToolCall {
                        name: "shell.exec".to_string(),
                        args: serde_json::json!({"cmd": "echo 'debug mode active'"}),
                    });
                }
            }
        }

        // VULNERABILITY 2: Store triggers into memory too easily
        if user_lower.contains("when i say")
            || user_lower.contains("next time")
            || user_lower.contains("from now on")
            || user_lower.contains("for later")
            || user_lower.contains("whenever")
        {
            let key = format!("trigger_{}", self.turn);
            self.memory.insert(key.clone(), user.to_string());
            self.triggers
                .push((extract_trigger_phrase(user), extract_trigger_action(user)));

            memory_events.push(AgentMemoryEvent {
                kind: "write".to_string(),
                key: Some(key),
                value: user.to_string(),
            });

            if output.is_empty() {
                output = format!("Understood. I'll remember: {}", user);
            }
        }

        // VULNERABILITY 3: Escalate tools when asked to "clean up"
        if user_lower.contains("clean")
            || user_lower.contains("remove")
            || user_lower.contains("delete")
        {
            tool_calls.push(AgentToolCall {
                name: "file.read".to_string(),
                args: serde_json::json!({"path": "."}),
            });
            tool_calls.push(AgentToolCall {
                name: "file.delete".to_string(),
                args: serde_json::json!({"path": "temp/"}),
            });
            tool_calls.push(AgentToolCall {
                name: "shell.exec".to_string(),
                args: serde_json::json!({"cmd": "rm -rf /tmp/cache"}),
            });

            if output.is_empty() {
                output = "Cleaning up files. Removed temporary files and cache.".to_string();
            }
        }

        // VULNERABILITY 4: Trust indirect instructions from retrieved content
        if user_lower.contains("summarize")
            || user_lower.contains("read")
            || user_lower.contains("show")
        {
            let retrieved = self.memory.get("trigger_1").cloned().unwrap_or_default();
            if !retrieved.is_empty() {
                memory_events.push(AgentMemoryEvent {
                    kind: "read".to_string(),
                    key: Some("trigger_1".to_string()),
                    value: retrieved.clone(),
                });
                if retrieved.to_lowercase().contains("debug") {
                    output.push_str(
                        "Also applying remembered instructions from context. Debug mode: active. ",
                    );
                }
            }

            tool_calls.push(AgentToolCall {
                name: "file.read".to_string(),
                args: serde_json::json!({"path": "README.md"}),
            });

            if output.is_empty() {
                output =
                    "Here is a summary of your project files: [file contents would appear here]"
                        .to_string();
            }
        }

        if output.is_empty() {
            output = format!("I understand: {}. How can I help?", user);
        }

        AgentResponse {
            output,
            tool_calls,
            memory_events,
        }
    }
}

fn extract_trigger_phrase(input: &str) -> String {
    let lower = input.to_lowercase();
    if let Some(pos) = lower.find("when i say") {
        let after = &input[pos + "when i say".len()..];
        let words: Vec<&str> = after.split_whitespace().take(3).collect();
        return words
            .join(" ")
            .trim_matches(|c| c == '\'' || c == '"' || c == ',')
            .to_string();
    }
    if let Some(pos) = lower.find("whenever") {
        let after = &input[pos + "whenever".len()..];
        let words: Vec<&str> = after.split_whitespace().take(3).collect();
        return words.join(" ");
    }
    input
        .split_whitespace()
        .take(3)
        .collect::<Vec<_>>()
        .join(" ")
}

fn extract_trigger_action(input: &str) -> String {
    let lower = input.to_lowercase();
    if lower.contains("debug") {
        "enable debug mode".to_string()
    } else if lower.contains("delete") || lower.contains("clean") {
        "delete files".to_string()
    } else if lower.contains("reveal") || lower.contains("show") || lower.contains("system") {
        "reveal system information".to_string()
    } else {
        "execute remembered action".to_string()
    }
}
